File: lab5-collections/python/playlistColls.py
class Piece:
    def __init__(self, name, performer, length_secs): # constructor
        self.name = name
        self.performer = performer
        self.length_secs = length_secs
        self.check()

    def __repr__(self):  # formatting (like toString())
        result = "%s by %s (%.2fs)" % (self.name, self.performer, self.length_secs)
        return result

    def check(self):
        # do some dynamic type checking:
        assert self.name.__class__ == str
        assert self.performer.__class__ == str
        assert self.length_secs.__class__ == type(1.0)
        l = self.length_secs
        assert 0 < l and l < 36000

class Product:
    def __init__(self, name, brand):
        self.name = name
        self.brand = brand
        self.check()

    def __repr__(self):  # formatting (like toString())
        result = "%s by %s" % (self.name, self.brand)
        return result

    def check(self):
        # do some dynamic type checking:
        assert self.name.__class__ == str
        assert self.brand.__class__ == str

class Advert:
    def __init__(self, product, length_secs): # constructor
        self.product = product
        self.length_secs = length_secs
        self.check()

    def __repr__(self):  # formatting (like toString())
        result = "Advert for %s by %s (%.2fs)" % (self.product.name, self.product.brand, self.length_secs)
        return result

    def check(self):
        # do some dynamic type checking:
        assert self.product.__class__ == Product
        assert self.length_secs.__class__ == float
        l = self.length_secs
        assert 0 < l and l < 120

piece1 = Piece("Moonlight", "C. Arrau", 17*60+26.0)
piece2 = Piece("Pathetique", "D. Barenboim", 16*60+49.0)
advert1 = Advert(Product("Bounty", "Mars"), 15.0)

# a heterogeneous list:
things = [111, piece1, "hello", advert1, [piece2]]

print("things = %s" % things)

piecesFromThings = "todo" #TASK 5.3.(c) -- replace "todo" by a list comprehension

print("piecesFromThings = %s" % piecesFromThings)

playlist1 = [piece1, advert1, piece2]

print("playlist1 = %s" % playlist1)

lengths1 = [ item.length_secs for item in playlist1 ]

print("lengths1 = %s" % lengths1)

playlist1noAds = [ item for item in playlist1 if item.__class__ != Advert ] 

print("playlist1noAds = %s" % playlist1noAds)

shortItemLenghts1 = "todo" # TASK 5.3.(b) -- replace "todo" by a list comprehension

print("shortItemLenghts1 = %s" % shortItemLenghts1)

# empty line
print

# heterogeneous dictionary:
pieceToScore = { piece1 : 10, piece2 : "dunno" }
print("pieceToScore = %s" % pieceToScore)
print ("piece1's score = %s" % pieceToScore[ piece1 ]) # like array lookup
print ("piece2's score = %s" % pieceToScore[ piece2 ])
pieceToScore[ piece2 ] = 10 # like array update
print("(after update) pieceToScore = %s" % pieceToScore)

File: lab5-collections/haskell/playlistColls.hs
module Main where

import Text.Printf

data Product
    = Product
        {
          product_name :: String
        , product_brand :: String
        }
    deriving (Eq)

instance Show Product where 
    show (Product name brand) = 
        printf "%s by %s" name brand

data Item
    = Piece
        {
            item_name :: String,
            item_performer :: String,
            item_length_secs :: Float
        }
    | Advert
        {
            item_product :: Product,
            item_length_secs :: Float
        }
    deriving (Eq)

isAdvert (Advert _ _) = True
isAdvert _ = False

itemOK (Piece _ _ len) = 0 < len && len < 36000
itemOK (Advert _ len) = 0 < len && len < 120

instance (Show Item) where
    show (Piece name performer len) =
        printf "%s by %s (%.1fs)" name performer len
    show (Advert product len) =
        printf "Advert for %s (%.1fs)" (show product) len

piece1 = Piece "Moonlight" "C. Arrau" (17*60+26)
piece2 = Piece "Pathetique" "D. Barenboim" (16*60+49)
advert1 = Advert (Product "Bounty" "Mars") 15

playlist1 = [piece1, advert1, piece2]

lengths1 = [ item_length_secs item | item <- playlist1 ]

playlist1noAds = [ item | item <- playlist1, not (isAdvert item) ]

playlist2 = [piece2, advert1]

playlists = [playlist1, playlist2] -- a list of lists

adsFromPlaylists = "todo" -- TASK 5.4(b)

shortItemLenghts1 = "todo" -- TASK 5.4(a)

main =
    do
    printf "playlist1 = %s\n" (show playlist1)
    printf "lenghts1 = %s\n" (show lengths1)
    printf "playlist1noAds = %s\n" (show playlist1noAds)
    printf "shortItemLenghts1 = %s\n" (show shortItemLenghts1)
    putStrLn ""
    printf "playlist2 = %s\n" (show playlist2)
    printf "playlists = %s\n" (show playlists)
    printf "adsFromPlaylists = %s\n" (show adsFromPlaylists)

